fairscale/optim/oss.py

```diff
@@ -175,25 +197,63 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
             loss = self.optim.step(**kwargs)
 
         # Sync all the states. Broadcast requests are issued async, we check completeness before moving on
-        requests = []
-        requires_grad = []
-        for rank, param_groups in enumerate(self.partition_parameters()):
-            for param_group in param_groups:
-                for param in param_group["params"]:
-                    # NOTE: Broadcast is in-place and not differentiable
-                    # Gloo will rightly assert on this operation for any tensor that requires grad.
-                    # We save and restore the grad requirement state to work around that, in our case
-                    # the grad is only useful on the source rank.
-                    global_rank = self.get_global_rank(self.group, rank)
+        for device_params in self.per_device_params:  # all the params on this device (inc all ranks)
+            self._broadcast_params_task(
+                self._buffers[device_params[0][0].device], device_params, self.group, self.global_rank
+            )
+
+        return loss
+
+    @staticmethod
+    def _broadcast_params_task(buffer: torch.Tensor, params: List[List[Parameter]], group: Any, self_rank: int) -> None:
+        """Helper function to broadcast all the parameters
+        """
+        buffer_size = buffer.numel()
+
+        for rank, rank_params in enumerate(params):  # all the params sorted per rank
+            restore_require_grad = []
 
-                    requires_grad.append((param, param.requires_grad))
+            if len(rank_params) == 0:
+                continue
+
+            global_rank = OSS.get_global_rank(group, rank)
+
+            # Copy small gradients into per-GPU buffers and then async broadcast
+            i_p = 0
+            offset = 0
+            while i_p < len(rank_params) and offset + rank_params[i_p].numel() < buffer_size:
+                end = offset + rank_params[i_p].numel()
+                if rank == self_rank:
+                    buffer[offset:end].copy_(rank_params[i_p].data.view(-1))  # type: ignore
+                offset = end
+                i_p += 1
+
+            if i_p > 0:
+                dist.broadcast(tensor=buffer, src=global_rank, group=group, async_op=True)
```

Comment:
corner case: this could be the last param, which means async_op needs to == True?

Reply:
good point, that could be what the unit test is reacting to, fixing that !